src/buffer.rs

```diff
@@ -0,0 +1,294 @@
+use crate::ctx::TryIntoCtx;
+use crate::{Endian, Pwrite};
+use core::ops::{Index, IndexMut};
+
+/// A buffer designed for `Pwrite` that dynamically expands to hold all types written into it.
+///
+/// # Use Cases
+///
+/// * You don't know exactly how much you're writing ahead of time.
+/// * Your types change size depending on context parameters or specific values.
+///
+/// # Allocations
+///
+/// By default, this buffer expands every time one byte is written. The `with_increment`
+/// constructor can be used to change how much the buffer expands to reduce overall number of allocations.
+///
+/// # Pwrite and TryIntoCtx
+///
+/// Types must implement `TryIntoCtx` with a second generic parameter of `DynamicBuffer` to be used.
+///
+/// ```rust
+/// use scroll::{ctx::TryIntoCtx, DynamicBuffer, Error, LE, Pwrite};
+///
+/// struct EvenOdd(u32);
+///
+/// impl TryIntoCtx<(), DynamicBuffer> for EvenOdd {
+///     type Error = Error;
+///
+///     fn try_into_ctx(self, buf: &mut DynamicBuffer, _: ()) -> Result<usize, Self::Error> {
+///         let offset = &mut 0;
+///
+///         if self.0 % 2 == 0 {
+///             buf.gwrite_with(self.0, offset, LE)?;
+///         } else {
+///             buf.gwrite_with(self.0 as u64, offset, LE)?;
+///         }
+///
+///         Ok(*offset)
+///     }
+/// }
+///
+/// let mut buf = DynamicBuffer::new();
+///
+/// let offset = buf.pwrite(EvenOdd(2), 0).unwrap();
+/// assert_eq!(buf.get(), [2, 0, 0, 0]);
+/// buf.pwrite(EvenOdd(3), offset).unwrap();
+/// assert_eq!(buf.get()[offset..], [3, 0, 0, 0, 0, 0, 0, 0]);
+/// ```
+///
+/// ## Why can't I use types that already implement the default TryIntoCtx?
+///
+/// The default type parameters for `TryIntoCtx` expose a mutable raw slice to write into.
+/// This is great for pre-allocated buffers and types that have a static written size; however,
+/// we cannot create a custom slice type that performs dynamic expansions under the hood.
+///
+/// As a result, we need to expose a separate writing type that can track offsets and calculate when to allocate.
+///
+/// However, if your `TryIntoCtx` impls don't use any special slice APIs and just use `Pwrite` and/or
+/// basic indexing, it's extremely easy to migrate! Just add the `DynamicBuffer` generic type.
+pub struct DynamicBuffer {
+    buffer: Vec<u8>,
+    alloc_increment: usize,
+    start_offset: usize,
```

Comment:
reading through this it isn't exactly clear why start_offset is for, and why the extra bookkeeping associated with it; i'm sure it is required, but might be nice to have some comments explaining it's use, etc.